File: plugin/android/src/main/kotlin/ai/passio/nutrition_ai/converter/OutputConverter.kt
package ai.passio.nutrition_ai.converter

import ai.passio.nutrition_ai.R
import ai.passio.nutrition_ai.utils.toByteArray
import ai.passio.passiosdk.core.config.PassioMode
import ai.passio.passiosdk.core.config.PassioSDKError
import ai.passio.passiosdk.core.config.PassioStatus
import ai.passio.passiosdk.passiofood.BarcodeCandidate
import ai.passio.passiosdk.passiofood.DetectedCandidate
import ai.passio.passiosdk.passiofood.FoodCandidates
import ai.passio.passiosdk.passiofood.PackagedFoodCandidate
import ai.passio.passiosdk.passiofood.PassioID
import ai.passio.passiosdk.passiofood.PassioNutrient
import ai.passio.passiosdk.passiofood.data.measurement.Grams
import ai.passio.passiosdk.passiofood.data.measurement.KiloCalories
import ai.passio.passiosdk.passiofood.data.measurement.Kilograms
import ai.passio.passiosdk.passiofood.data.measurement.Micrograms
import ai.passio.passiosdk.passiofood.data.measurement.Milligrams
import ai.passio.passiosdk.passiofood.data.measurement.Milliliters
import ai.passio.passiosdk.passiofood.data.measurement.Unit
import ai.passio.passiosdk.passiofood.data.measurement.UnitEnergy
import ai.passio.passiosdk.passiofood.data.measurement.UnitMass
import ai.passio.passiosdk.passiofood.data.model.PassioAlternative
import ai.passio.passiosdk.passiofood.data.model.PassioFoodItemData
import ai.passio.passiosdk.passiofood.data.model.PassioFoodOrigin
import ai.passio.passiosdk.passiofood.data.model.PassioFoodRecipe
import ai.passio.passiosdk.passiofood.data.model.PassioIDAttributes
import ai.passio.passiosdk.passiofood.data.model.PassioIDEntityType
import ai.passio.passiosdk.passiofood.data.model.PassioServingSize
import ai.passio.passiosdk.passiofood.data.model.PassioServingUnit
import android.content.Context
import android.graphics.Bitmap
import android.net.Uri

fun mapFromPassioStatus(status: PassioStatus): Map<String, Any?> {
    val map = mutableMapOf<String, Any?>()
    map["debugMessage"] = status.debugMessage
    map["activeModels"] = status.activeModels
    map["missingFiles"] = status.missingFiles
    map["mode"] = status.mode.toPlatformString()
    map["error"] = status.error?.toPlatformString()
    return map
}

private fun PassioMode.toPlatformString(): String {
    return when (this) {
        PassioMode.NOT_READY -> "notReady"
        PassioMode.FAILED_TO_CONFIGURE -> "failedToConfigure"
        PassioMode.IS_BEING_CONFIGURED -> "isBeingConfigured"
        PassioMode.IS_DOWNLOADING_MODELS -> "isDownloadingModels"
        PassioMode.IS_READY_FOR_DETECTION -> "isReadyForDetection"
    }
}

private fun PassioSDKError.toPlatformString(): String {
    return when (this) {
        PassioSDKError.MIN_SDK_VERSION -> "minSDKVersion"
        PassioSDKError.KEY_NOT_VALID -> "keyNotValid"
        PassioSDKError.MODELS_NOT_VALID -> "modelsNotValid"
        PassioSDKError.LICENSE_KEY_HAS_EXPIRED -> "licensedKeyHasExpired"
        PassioSDKError.LICENSE_DECODING_ERROR -> "licenseDecodingError"
        PassioSDKError.MODELS_DOWNLOAD_FAILED -> "modelsDownloadFailed"
        PassioSDKError.NO_INTERNET_CONNECTION -> "noInternetConnection"
        PassioSDKError.NO_MODELS_FILES_FOUND -> "noModelsFilesFound"
        PassioSDKError.NOT_LICENSED_FOR_THIS_PROJECT -> "notLicensedForThisProject"
        PassioSDKError.MISSING_DEPENDENCY -> "missingDependency"
    }
}

fun mapFromFoodCandidates(candidates: FoodCandidates): Map<String, Any?> {
    val candidatesMap = mutableMapOf<String, Any?>()
    candidatesMap["detectedCandidate"] = candidates.detectedCandidates?.map {
        mapFromDetectedCandidate(it)
    }
    candidatesMap["barcodeCandidates"] = candidates.barcodeCandidates?.map {
        mapFromBarcodeCandidate(it)
    }
    candidatesMap["packagedFoodCandidates"] = candidates.packagedFoodCandidates?.map {
        mapFromPackagedFoodCode(it)
    }
    return candidatesMap
}

fun mapFromPackagedFoodCode(candidate: PackagedFoodCandidate): Map<String, Any?> {
    val candidateMap = mutableMapOf<String, Any?>()
    candidateMap["packagedFoodCode"] = candidate.packagedFoodCode
    candidateMap["confidence"] = candidate.confidence
    return candidateMap
}

fun mapFromBarcodeCandidate(candidate: BarcodeCandidate): Map<String, Any?> {
    val candidateMap = mutableMapOf<String, Any?>()
    candidateMap["value"] = candidate.barcode
    candidateMap["boundingBox"] = arrayListOf(
        candidate.boundingBox.left,
        candidate.boundingBox.top,
        candidate.boundingBox.width(),
        candidate.boundingBox.height(),
    )
    return candidateMap
}

private fun mapFromDetectedCandidate(candidate: DetectedCandidate): Map<String, Any?> {
    val candidateMap = mutableMapOf<String, Any?>()
    candidateMap["passioID"] = candidate.passioID
    candidateMap["confidence"] = candidate.confidence
    candidateMap["boundingBox"] = arrayListOf(
        candidate.boundingBox.left,
        candidate.boundingBox.top,
        candidate.boundingBox.width(),
        candidate.boundingBox.height(),
    )
    return candidateMap
}

fun mapFromAlternatives(alternative: PassioAlternative): Map<String, Any?> {
    val alternativeMap = mutableMapOf<String, Any?>()
    alternativeMap["passioID"] = alternative.passioID
    alternativeMap["name"] = alternative.name
    alternativeMap["quantity"] = alternative.number
    alternativeMap["unitName"] = alternative.unit
    return alternativeMap
}

fun mapFromServingSize(servingSize: PassioServingSize): Map<String, Any?> {
    val servingSizeMap = mutableMapOf<String, Any?>()
    servingSizeMap["quantity"] = servingSize.quantity
    servingSizeMap["unitName"] = servingSize.unitName
    return servingSizeMap
}

fun mapFromServingUnit(servingUnit: PassioServingUnit): Map<String, Any?> {
    val servingUnitMap = mutableMapOf<String, Any?>()
    servingUnitMap["unitName"] = servingUnit.unitName
    servingUnitMap["weight"] = mapFromUnitMass(servingUnit.weight)
    return servingUnitMap
}

private fun mapFromUnitMass(unitMass: UnitMass?): Map<String, Any?>? {
    if (unitMass == null) return null
    val unitMassMap = mutableMapOf<String, Any?>()
    unitMassMap["unit"] = unitMass.unit.unitString()
    unitMassMap["value"] = unitMass.value
    return unitMassMap
}

private fun mapFromUnitEnergy(unitEnergy: UnitEnergy?): Map<String, Any?>? {
    if (unitEnergy == null) return null
    val unitEnergyMap = mutableMapOf<String, Any?>()
    unitEnergyMap["unit"] = unitEnergy.unit.unitString()
    unitEnergyMap["value"] = unitEnergy.value
    return unitEnergyMap
}

private fun mapFromUnitIU(value: Double?): Map<String, Any?>? {
    if (value == null) return null
    val unitEnergyMap = mutableMapOf<String, Any?>()
    unitEnergyMap["unit"] = "iu"
    unitEnergyMap["value"] = value
    return unitEnergyMap
}

private fun Unit.unitString(): String {
    return when (this) {
        is Kilograms -> "kilograms"
        is Grams -> "grams"
        is Milligrams -> "milligrams"
        is Micrograms -> "micrograms"
        is KiloCalories -> "kilocalories"
        is Milliliters -> "grams"
        else -> throw java.lang.IllegalArgumentException("No known unit: $this")
    }
}

fun mapFromFoodOrigin(origin: PassioFoodOrigin): Map<String, Any?> {
    val originMap = mutableMapOf<String, Any?>()
    originMap["id"] = origin.id
    originMap["source"] = origin.source
    originMap["licenseCopy"] = origin.licenseCopy
    return originMap
}

fun mapFromFoodItemData(data: PassioFoodItemData): Map<String, Any?> {
    val dataMap = mutableMapOf<String, Any?>()
    val originalQuantity = data.selectedQuantity
    val originalUnit = data.selectedUnit
    dataMap["passioID"] = data.passioID
    dataMap["name"] = data.name
    dataMap["selectedQuantity"] = data.selectedQuantity
    dataMap["selectedUnit"] = data.selectedUnit
    dataMap["entityType"] = data.entityType.name
    dataMap["servingUnits"] = data.servingUnits.map { mapFromServingUnit(it) }
    dataMap["servingSizes"] = data.servingSizes.map { mapFromServingSize(it) }
    dataMap["ingredientsDescription"] = data.ingredientsDescription
    dataMap["barcode"] = data.barcode
    dataMap["foodOrigins"] = data.foodOrigins?.map { mapFromFoodOrigin(it) }
    dataMap["referenceWeight"] = mapFromUnitMass(data.referenceWeight)
    dataMap["parents"] = data.parents?.map { mapFromAlternatives(it) }
    dataMap["children"] = data.children?.map { mapFromAlternatives(it) }
    dataMap["siblings"] = data.siblings?.map { mapFromAlternatives(it) }
    dataMap["tags"] = data.tags
    // Map the nutrient per 100 grams
    data.setServingSize("gram", 100.0)
    dataMap["calories"] = mapFromUnitEnergy(data.totalCalories())
    dataMap["carbs"] = mapFromUnitMass(data.totalCarbs())
    dataMap["fat"] = mapFromUnitMass(data.totalFat())
    dataMap["proteins"] = mapFromUnitMass(data.totalProtein())
    dataMap["saturatedFat"] = mapFromUnitMass(data.totalSatFat())
    dataMap["transFat"] = mapFromUnitMass(data.totalTransFat())
    dataMap["monounsaturatedFat"] = mapFromUnitMass(data.totalMonounsaturatedFat())
    dataMap["polyunsaturatedFat"] = mapFromUnitMass(data.totalPolyunsaturatedFat())
    dataMap["cholesterol"] = mapFromUnitMass(data.totalCholesterol())
    dataMap["sodium"] = mapFromUnitMass(data.totalSodium())
    dataMap["fibers"] = mapFromUnitMass(data.totalFibers())
    dataMap["sugars"] = mapFromUnitMass(data.totalSugars())
    dataMap["sugarsAdded"] = mapFromUnitMass(data.totalSugarsAdded())
    dataMap["vitaminD"] = mapFromUnitMass(data.totalVitaminD())
    dataMap["calcium"] = mapFromUnitMass(data.totalCalcium())
    dataMap["iron"] = mapFromUnitMass(data.totalIron())
    dataMap["potassium"] = mapFromUnitMass(data.totalPotassium())
    dataMap["vitaminA"] = mapFromUnitIU(data.totalVitaminA())
    dataMap["vitaminC"] = mapFromUnitMass(data.totalVitaminC())
    dataMap["alcohol"] = mapFromUnitMass(data.totalAlcohol())
    dataMap["sugarAlcohol"] = mapFromUnitMass(data.totalSugarAlcohol())
    dataMap["vitaminB12Added"] = mapFromUnitMass(data.totalVitaminB12Added())
    dataMap["vitaminB12"] = mapFromUnitMass(data.totalVitaminB12())
    dataMap["vitaminB6"] = mapFromUnitMass(data.totalVitaminB6())
    dataMap["vitaminE"] = mapFromUnitMass(data.totalVitaminE())
    dataMap["vitaminEAdded"] = mapFromUnitMass(data.totalVitaminEAdded())
    dataMap["magnesium"] = mapFromUnitMass(data.totalMagnesium())
    dataMap["phosphorus"] = mapFromUnitMass(data.totalPhosphorus())
    dataMap["iodine"] = mapFromUnitMass(data.totalIodine())
    data.setServingSize(originalUnit, originalQuantity)
    return dataMap
}

fun mapFromPassioIDAttributes(attrs: PassioIDAttributes): Map<String, Any?> {
    val attrMap = mutableMapOf<String, Any?>()
    attrMap["passioID"] = attrs.passioID
    attrMap["name"] = attrs.name
    attrMap["entityType"] = attrs.entityType.name
    attrMap["foodItem"] = if (attrs.passioFoodItemData != null) mapFromFoodItemData(attrs.passioFoodItemData!!) else null
    attrMap["recipe"] = if (attrs.passioFoodRecipe != null) mapFromPassioFoodRecipe(attrs.passioFoodRecipe!!) else null
    attrMap["parents"] = attrs.parents?.map { mapFromAlternatives(it) }
    attrMap["siblings"] = attrs.siblings?.map { mapFromAlternatives(it) }
    attrMap["children"] = attrs.children?.map { mapFromAlternatives(it) }
    return attrMap
}

fun mapFromPassioFoodRecipe(recipe: PassioFoodRecipe): Map<String, Any?> {
    val recipeMap = mutableMapOf<String, Any?>()
    recipeMap["passioID"] = recipe.passioID
    recipeMap["name"] = recipe.name
    recipeMap["selectedQuantity"] = recipe.selectedQuantity
    recipeMap["selectedUnit"] = recipe.selectedUnit
    recipeMap["servingUnits"] = recipe.servingUnits.map { mapFromServingUnit(it) }
    recipeMap["servingSizes"] = recipe.servingSizes.map { mapFromServingSize(it) }
    recipeMap["foodItems"] = recipe.foodItems.map { mapFromFoodItemData(it) }
    return recipeMap
}

fun mapFromSearchResult(result: Pair<PassioID, String>): Map<String, Any?> {
    val searchMap = mutableMapOf<String, Any?>()
    searchMap["passioID"] = result.first
    searchMap["name"] = result.second
    return searchMap
}

fun mapFromBitmap(image: Bitmap?): MutableMap<String, Any?>? {
    // Checking if the 'image' is not null before performing operations
    val byteArray: ByteArray? = image?.toByteArray()

    // Creating a map from image and byteArray properties if 'image' is not null
    return image?.let {
        // Initializing a mutable map to store image properties
        val map = mutableMapOf<String, Any?>()

        // Adding width, height, and pixel data to the map
        map["width"] = it.width
        map["height"] = it.height
        map["pixels"] = byteArray!!

        // The resulting map containing image properties
        map
    }
}

/// Maps the information from a completed downloading file to a [Map].
///
/// Parameters:
/// - [fileUri]: The URI of the completed file.
/// - [filesLeft]: The number of files left to download.
///
/// Returns a [Map] containing the mapped information.
fun mapFromCompletedDownloadingFile(fileUri: Uri, filesLeft: Int): Map<String, Any?> {
    val downloadMap = mutableMapOf<String, Any?>()
    downloadMap["fileUri"] = fileUri.path
    downloadMap["filesLeft"] = filesLeft
    return downloadMap
}

/**
 * This function takes an `event` and `data` as parameters and maps them into a
 * `Map` representing the structure of PassioStatusListener events.
 *
 * @param event The type of event (e.g., "onPassioStatusChanged").
 * @param data The data associated with the event. It can be of any type.
 * @return A `Map` containing the event type and associated data.
 */
fun mapFromPassioStatusListener(event: String, data: Any?): Map<String, Any?> {
    val statusMap = mutableMapOf<String, Any?>()
    statusMap["event"] = event
    statusMap["data"] = data
    return statusMap
}

/**
 * Maps a [PassioNutrient] object to a [Map] for serialization.
 *
 * @param nutrient The [PassioNutrient] object to be mapped.
 * @return A [Map] containing the serialized representation of the [PassioNutrient].
 */
fun mapFromPassioNutrient(nutrient: PassioNutrient): Map<String, Any?> {
    // Create a mutable map to store the serialized nutrient data.
    val nutrientMap = mutableMapOf<String, Any?>()

    // Add nutrient properties to the map.
    nutrientMap["amount"] = nutrient.amount
    nutrientMap["inflammatoryEffectScore"] = nutrient.inflammatoryEffectScore
    nutrientMap["name"] = nutrient.name
    nutrientMap["unit"] = nutrient.unit

    // Return the serialized nutrient map.
    return nutrientMap
}

File: plugin/ios/Classes/convertor/OutputConverter.swift
import PassioNutritionAISDK

struct OutputConverter {

 func mapFromFoodCandidates(candidates: FoodCandidates) -> [String: Any?] {
        var candidatesMap: [String: Any?] = [:]
        
        let detectedCandidates = candidates.detectedCandidates
        let mappedCandidates = detectedCandidates.map {
            mapFromDetectedCandidate(detectedCandidate: $0)
        }
        candidatesMap["detectedCandidate"] = mappedCandidates
        
        if let barcodeCandidates = candidates.barcodeCandidates {
            let mappedCandidates = barcodeCandidates.map {
                mapFromBarcodeCandidate(barcodeCandidate: $0)
            }
            candidatesMap["barcodeCandidates"] = mappedCandidates
        }
        
        if let packagedFoodCandidates = candidates.packagedFoodCandidates {
            let mappedCandidates = packagedFoodCandidates.map {
                mapFromPackagedFoodCandidate(packagedFoodCandidate: $0)
            }
            candidatesMap["packagedFoodCandidates"] = mappedCandidates
        }
        return candidatesMap
    }
    
    func mapFromDetectedCandidate(detectedCandidate: DetectedCandidate) -> [String: Any?] {
        var candidateMap: [String: Any?] = [:]
        candidateMap["passioID"] = detectedCandidate.passioID
        candidateMap["confidence"] = detectedCandidate.confidence
        let boundingBox = detectedCandidate.boundingBox
        candidateMap["boundingBox"] = mapFromBoundingBox(boundingBox: boundingBox)
        if let amountEstimate = detectedCandidate.amountEstimate {
            candidateMap["amountEstimate"] = mapAmounteEstimage(amountEstimate: amountEstimate )
        }
        return candidateMap
    }
    
    func mapFromBarcodeCandidate(barcodeCandidate: BarcodeCandidate) -> [String: Any?] {
        var candidateMap: [String: Any?] = [:]
        candidateMap["value"] = barcodeCandidate.value
        let boundingBox = barcodeCandidate.boundingBox
        candidateMap["boundingBox"] = mapFromBoundingBox(boundingBox: boundingBox)
        return candidateMap
    }
    
    func mapFromBoundingBox(boundingBox: CGRect) -> [Double] { 
    return [
            boundingBox.origin.x,
            boundingBox.origin.y,
            boundingBox.width,
            boundingBox.height,
        ]
    }

    func mapFromPackagedFoodCandidate(packagedFoodCandidate: PackagedFoodCandidate) -> [String: Any?] {
        var candidateMap: [String: Any?] = [:]
        candidateMap["packagedFoodCode"] = packagedFoodCandidate.packagedFoodCode
        candidateMap["confidence"] = packagedFoodCandidate.confidence
        return candidateMap
    }
    
    
    func mapFromPassioIDAttributes(passioIDAttributes: PassioIDAttributes) -> [String: Any?] {
        var attrMap = [String: Any?]()
        attrMap["passioID"] = passioIDAttributes.passioID
        attrMap["name"] = passioIDAttributes.name
        attrMap["entityType"] = passioIDAttributes.entityType.rawValue
        attrMap["foodItem"] = passioIDAttributes.passioFoodItemData != nil ?
        mapFromFoodItemData(itemData: passioIDAttributes.passioFoodItemData!) : nil
        attrMap["recipe"] = passioIDAttributes.recipe != nil ?
        mapFromPassioFoodRecipe(recipe: passioIDAttributes.recipe!) : nil
        attrMap["parents"] = passioIDAttributes.parents?.map { mapFromAlternatives($0) }
        attrMap["siblings"] = passioIDAttributes.siblings?.map { mapFromAlternatives($0) }
        attrMap["children"] = passioIDAttributes.children?.map { mapFromAlternatives($0) }
        return attrMap
    }
    
    func mapFromFoodItemData(itemData: PassioFoodItemData) -> [String: Any?] {
        var data = itemData
        var dataMap = [String: Any?]()
        dataMap["passioID"] = data.passioID
        dataMap["name"] = data.name
        dataMap["selectedQuantity"] = data.selectedQuantity
        dataMap["selectedUnit"] = data.selectedUnit
        dataMap["entityType"] = data.entityType.rawValue
        dataMap["servingUnits"] = data.servingUnits.map { mapFromServingUnit($0) }
        dataMap["servingSizes"] = data.servingSizes.map { mapFromServingSize($0) }
        dataMap["ingredientsDescription"] = data.ingredientsDescription
        dataMap["barcode"] = data.barcode
        dataMap["foodOrigins"] = data.foodOrigins?.map { mapFromFoodOrigin($0) } ?? []
        dataMap["parents"] = data.parents?.map { mapFromAlternatives($0) }
        dataMap["children"] = data.children?.map { mapFromAlternatives($0) }
        dataMap["siblings"] = data.siblings?.map { mapFromAlternatives($0) }
        dataMap["tags"] = data.tags?.map { $0 }
        // Map the nutrient per 100 grams
        _ = data.setFoodItemDataServingSize(unit: "gram", quantity: 100)
        dataMap["calories"] = mapFromUnitEnergy(data.totalCalories)
        dataMap["carbs"] = mapFromUnitMass(data.totalCarbs)
        dataMap["fat"] = mapFromUnitMass(data.totalFat)
        dataMap["proteins"] = mapFromUnitMass(data.totalProteins)
        dataMap["saturatedFat"] = mapFromUnitMass(data.totalSaturatedFat)
        dataMap["transFat"] = mapFromUnitMass(data.totalTransFat)
        dataMap["monounsaturatedFat"] = mapFromUnitMass(data.totalMonounsaturatedFat)
        dataMap["polyunsaturatedFat"] = mapFromUnitMass(data.totalPolyunsaturatedFat)
        dataMap["cholesterol"] = mapFromUnitMass(data.totalCholesterol)
        dataMap["sodium"] = mapFromUnitMass(data.totalSodium)
        dataMap["fibers"] = mapFromUnitMass(data.totalFibers)
        dataMap["sugars"] = mapFromUnitMass(data.totalSugars)
        dataMap["sugarsAdded"] = mapFromUnitMass(data.totalSugarsAdded)
        dataMap["vitaminD"] = mapFromUnitMass(data.totalVitaminD)
        dataMap["calcium"] = mapFromUnitMass(data.totalCalcium)
        dataMap["iron"] = mapFromUnitMass(data.totalIron)
        dataMap["potassium"] = mapFromUnitMass(data.totalPotassium)
        dataMap["vitaminA"] = mapFromUnitIU(data.totalVitaminA?.value)
        dataMap["vitaminC"] = mapFromUnitMass(data.totalVitaminC)
        dataMap["alcohol"] = mapFromUnitMass(data.totalAlcohol)
        dataMap["sugarAlcohol"] = mapFromUnitMass(data.totalSugarAlcohol)
        dataMap["vitaminB12Added"] = mapFromUnitMass(data.totalVitaminB12Added)
        dataMap["vitaminB12"] = mapFromUnitMass(data.totalVitaminB12)
        dataMap["vitaminB6"] = mapFromUnitMass(data.totalVitaminB6)
        dataMap["vitaminE"] = mapFromUnitMass(data.totalVitaminE)
        dataMap["vitaminEAdded"] = mapFromUnitMass(data.totalVitaminEAdded)
        dataMap["magnesium"] = mapFromUnitMass(data.totalMagnesium)
        dataMap["phosphorus"] = mapFromUnitMass(data.totalPhosphorus)
        dataMap["iodine"] = mapFromUnitMass(data.totalIodine)
        return dataMap
    }
    
    func mapFromPassioFoodRecipe(recipe: PassioFoodRecipe) -> [String: Any?] {
        var recipeMap = [String: Any?]()
        recipeMap["passioID"] = recipe.passioID
        recipeMap["name"] = recipe.name
        recipeMap["selectedQuantity"] = recipe.selectedQuantity
        recipeMap["selectedUnit"] = recipe.selectedUnit
        recipeMap["servingUnits"] = recipe.servingUnits.map { mapFromServingUnit($0) }
        recipeMap["servingSizes"] = recipe.servingSizes.map { mapFromServingSize($0) }
        recipeMap["foodItems"] = recipe.foodItems.map { mapFromFoodItemData(itemData: $0) }
        return recipeMap
    }
    
    func mapFromAlternatives(_ alternative: PassioAlternative) -> [String: Any?] {
        var alternativeMap = [String: Any?]()
        alternativeMap["passioID"] = alternative.passioID
        alternativeMap["name"] = alternative.name
        alternativeMap["quantity"] = alternative.quantity
        alternativeMap["unitName"] = alternative.unitName
        return alternativeMap
    }
    
    func mapFromServingSize(_ servingSize: PassioServingSize) -> [String: Any?] {
        var servingSizeMap = [String: Any?]()
        servingSizeMap["quantity"] = servingSize.quantity
        servingSizeMap["unitName"] = servingSize.unitName
        return servingSizeMap
    }
    
    func mapFromServingUnit(_ servingUnit: PassioServingUnit) -> [String: Any?] {
        var servingUnitMap = [String: Any?]()
        servingUnitMap["unitName"] = servingUnit.unitName
        servingUnitMap["weight"] = mapFromUnitMass(servingUnit.weight)
        return servingUnitMap
    }
    
    func mapFromFoodOrigin(_ origin: PassioFoodOrigin) -> [String: Any?] {
        var originMap = [String: Any?]()
        originMap["id"] = origin.id
        originMap["source"] = origin.source
        originMap["licenseCopy"] = origin.licenseCopy
        return originMap
    }
    
    func mapFromUnitEnergy(_ unitEnergy: Measurement<UnitEnergy>?) -> [String: Any?]? {
        guard let unitEnergy = unitEnergy else { return nil }
        var unitEnergyMap = [String: Any?]()
        unitEnergyMap["unit"] = convertEnergyUnit(unitEnergy: unitEnergy.unit)
        unitEnergyMap["value"] = unitEnergy.value
        return unitEnergyMap
    }
    
    private func mapFromUnitMass(_ unitMass: Measurement<UnitMass>?) -> [String: Any?]? {
        guard let unitMass = unitMass else { return nil }
        var unitMassMap = [String: Any?]()
        let unit = convertMashUnit(unitMass: unitMass.unit)
        unitMassMap["unit"] = unit
        unitMassMap["value"] = unitMass.value
        return unitMassMap
    }
    
    func mapFromUnitIU(_ value: Double?) -> [String: Any?]? {
        guard let value = value else { return nil }
        var unitEnergyMap = [String: Any?]()
        unitEnergyMap["unit"] = "iu"
        unitEnergyMap["value"] = value
        return unitEnergyMap
    }
    
    func convertEnergyUnit(unitEnergy: UnitEnergy) -> String {
        unitEnergy.symbol.description == "kCal" ? "kilocalories" : unitEnergy.symbol.description
    }
    
    func convertMashUnit(unitMass: UnitMass) -> String {
        switch unitMass.symbol.description {
        case "g","gr","ml":
            return "grams"
        case "kg":
            return "kilograms"
        case "mg":
            return "milligrams"
        case "µg":
            return "micrograms"
        default:
            return "grams"
        }
    }
    
    
    func mapAmounteEstimage(amountEstimate: AmountEstimate ) -> [String: Any?] {
        var mapEstimate = [String: Any?]()
        if let volumeEstimate = amountEstimate.volumeEstimate {
            mapEstimate["volumeEstimate"] = volumeEstimate
        }
        if let weightEstimate = amountEstimate.weightEstimate {
            mapEstimate["weightEstimate"] = weightEstimate
        }
        if let estimationQuality = amountEstimate.estimationQuality?.rawValue {
            mapEstimate["estimationQuality"] = estimationQuality
        }
        if let moveDevice = amountEstimate.moveDevice?.rawValue {
            mapEstimate["moveDevice"] = moveDevice
        }
        if let viewingAngle = amountEstimate.viewingAngle {
            mapEstimate["viewingAngle"] = viewingAngle
        }
        return mapEstimate
    }

    // MARK: - PassioStatus Mapping
    // This method maps PassioStatus properties to a dictionary for event reporting.
    func mapFromPassioStatus(passioStatus: PassioStatus) -> [String: Any?] {
        // Initialize an empty dictionary to store the mapped status information.
        var statusMap = [String: Any?]()
        // Map PassioStatus properties to corresponding dictionary keys.
        statusMap["mode"] = "\(passioStatus.mode)"
        statusMap["missingFiles"] = passioStatus.missingFiles
        statusMap["debugMessage"] = passioStatus.debugMessage
        statusMap["error"] = passioStatus.error != nil ? "\(passioStatus.error!)" : nil
        statusMap["activeModels"] = passioStatus.activeModels
        // Return the dictionary containing the mapped PassioStatus information.
        return statusMap
    }

    // MARK: - Completed Downloading File Mapping
    // This method maps completed downloading file details to a dictionary for event reporting.
    func mapFromCompletedDownloadingFile(fileUri: FileLocalURL, filesLeft: Int) -> [String: Any?] {
        // Initialize an empty dictionary to store the mapped download information.
        var downloadMap = [String: Any?]()
        // Map completed downloading file details to corresponding dictionary keys.
        downloadMap["fileUri"] = fileUri.absoluteString
        downloadMap["filesLeft"] = filesLeft
        // Return the dictionary containing the mapped download information.
        return downloadMap
    }

    // MARK: - PassioStatusListener Mapping
    // This method maps an event and associated data to a dictionary for event reporting.
    func mapFromPassioStatusListener(event: String, data: Any?) -> [String: Any?] {
        // Initialize an empty dictionary to store the mapped event and data.
        var statusListenerMap = [String: Any?]()
        // Map event and data to corresponding dictionary keys.
        statusListenerMap["event"] = event
        statusListenerMap["data"] = data
        // Return the dictionary containing the mapped event and data.
        return statusListenerMap
    }
    
    /**
     MARK: - Data Mapping

     Converts a PassioNutrient object into a dictionary.
     */
    func mapFromPassioNutrient(nutrient: PassioNutrient) -> [String: Any?] {
        // Create a mutable map to store the serialized nutrient data.
        var nutrientMap = [String: Any?]()

        // Add nutrient properties to the map.
        nutrientMap["amount"] = nutrient.amount
        nutrientMap["inflammatoryEffectScore"] = nutrient.inflammatoryEffectScore
        nutrientMap["name"] = nutrient.name
        nutrientMap["unit"] = nutrient.unit

        // Return the serialized nutrient map.
        return nutrientMap
    }
}
